cluster-autoscaler/core/scaledown/actuation/delete_in_batch.go

```diff
@@ -0,0 +1,187 @@
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+)
+
+const (
+	// MaxKubernetesEmptyNodeDeletionTime is the maximum time needed by Kubernetes to delete an empty node.
+	MaxKubernetesEmptyNodeDeletionTime = 3 * time.Minute
+	// MaxCloudProviderNodeDeletionTime is the maximum time needed by cloud provider to delete a node.
+	MaxCloudProviderNodeDeletionTime = 5 * time.Minute
+)
+
+// NodeBatcherDeleterImpl implements NodeBatcherDeleter interface.
+type NodeBatcherDeleterImpl struct {
+	sync.Mutex
+	ctx                   *context.AutoscalingContext
+	clusterState          *clusterstate.ClusterStateRegistry
+	nodeDeletionTracker   *deletiontracker.NodeDeletionTracker
+	deletionsPerNodeGroup map[string][]*apiv1.Node
+	deleteInterval        time.Duration
+	drainedNodeDeletions  map[string]bool
+}
+
+// NewNodeBatcherDeleter return new NodeBatchDeleter
+func NewNodeBatcherDeleter(ctx *context.AutoscalingContext, csr *clusterstate.ClusterStateRegistry, nodeDeletionTracker *deletiontracker.NodeDeletionTracker, deleteInterval time.Duration) deletionbatcher.NodeBatcherDeleter {
+	return &NodeBatcherDeleterImpl{
+		ctx:                   ctx,
+		clusterState:          csr,
+		nodeDeletionTracker:   nodeDeletionTracker,
+		deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+		deleteInterval:        deleteInterval,
+		drainedNodeDeletions:  make(map[string]bool),
+	}
+}
+
+// AddNode adds node to delete candidates and schedule deletion.
+func (d *NodeBatcherDeleterImpl) AddNode(node *apiv1.Node, drain bool) error {
+	nodeGroupId, first, err := d.addNodeToBucket(node, drain)
+	if err != nil {
+		return nil
+	}
+	if first {
+		go func(nodeGroupId string) {
+			time.Sleep(d.deleteInterval)
+			d.remove(nodeGroupId)
+		}(nodeGroupId)
+	}
+	return nil
+}
+
+// AddToBucket adds node to delete candidates and return if it's a first node in the group.
+func (d *NodeBatcherDeleterImpl) addNodeToBucket(node *apiv1.Node, drain bool) (string, bool, error) {
+	d.Lock()
+	defer d.Unlock()
+	nodeGroup, err := d.ctx.CloudProvider.NodeGroupForNode(node)
+	if err != nil {
+		return "", false, err
+	}
+	d.drainedNodeDeletions[node.Name] = drain
+	val, ok := d.deletionsPerNodeGroup[nodeGroup.Id()]
+	if !ok || len(val) == 0 {
+		d.deletionsPerNodeGroup[nodeGroup.Id()] = []*apiv1.Node{node}
+		return nodeGroup.Id(), true, nil
+	}
+	d.deletionsPerNodeGroup[nodeGroup.Id()] = append(d.deletionsPerNodeGroup[nodeGroup.Id()], node)
+	return nodeGroup.Id(), false, nil
+}
+
+// remove delete nodes of a given nodeGroup, if successful, the deletion is recorded in CSR, and an event is emitted on the node.
+func (d *NodeBatcherDeleterImpl) remove(nodeGroupId string) error {
+	d.Lock()
+	defer d.Unlock()
+	nodes, ok := d.deletionsPerNodeGroup[nodeGroupId]
+	if !ok {
+		return nil
+	}
+	delete(d.deletionsPerNodeGroup, nodeGroupId)
+
+	go func(nodes []*apiv1.Node, nodeGroupId string) {
+		var result status.NodeDeleteResult
+		nodeGroup, err := deleteNodesFromCloudProvider(d.ctx, nodes)
```

Comment:
I return nodeGroup, since recording ScaleDown events is happening outside of deleteNodesFromCloudProvider() function, however the function is exactly deleting nodes and nothing more. 
The name itself is quite long to add "andReturnNodeGroup", alternatively I can extract nodeGroup before calling deleteNodesFromCloudProvider() function and actually can pass nodeGroup to the function. Will it be better?